File: src/org/nextras/orm/intellij/completion/SetValueCompletionProvider.java
package org.nextras.orm.intellij.completion;


import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionProvider;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.psi.PsiElement;
import com.intellij.util.ProcessingContext;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty;
import com.jetbrains.php.lang.psi.elements.Field;
import com.jetbrains.php.lang.psi.elements.MethodReference;
import com.jetbrains.php.lang.psi.elements.ParameterList;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import org.jetbrains.annotations.NotNull;
import org.nextras.orm.intellij.utils.PhpIndexUtils;

public class SetValueCompletionProvider extends CompletionProvider<CompletionParameters> {

	@Override
	protected void addCompletions(@NotNull CompletionParameters parameters, ProcessingContext context, @NotNull CompletionResultSet result) {
		parameters.getPosition();
		PsiElement el = parameters.getPosition();
		if (
			el.getParent() == null
				|| el.getParent().getParent() == null
				|| !(el.getParent().getParent() instanceof ParameterList)
				|| ((ParameterList) el.getParent().getParent()).getParameters()[0] != el.getParent()
				|| el.getParent().getParent().getParent() == null
				|| !(el.getParent().getParent().getParent() instanceof MethodReference)
			) {
			return;
		}
		MethodReference methodReference = (MethodReference) el.getParent().getParent().getParent();
		if (!methodReference.getName().equals("setValue") && !methodReference.getName().equals("setReadOnlyValue")) {
			return;
		}
		PhpIndex phpIndex = PhpIndex.getInstance(el.getProject());
		for (PhpClass cls : PhpIndexUtils.getByType(methodReference.getClassReference().getType(), phpIndex)) {
			for (Field field : cls.getFields()) {
				if (!(field instanceof PhpDocProperty)) {
					continue;
				}
				result.addElement(LookupElementBuilder.create(field.getName()).withTypeText(field.getType().toString()));
			}
		}
	}
}

File: src/org/nextras/orm/intellij/completion/SetValueCompletionProvider.kt
package org.nextras.orm.intellij.completion

import com.intellij.codeInsight.completion.CompletionParameters
import com.intellij.codeInsight.completion.CompletionProvider
import com.intellij.codeInsight.completion.CompletionResultSet
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.util.ProcessingContext
import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty
import com.jetbrains.php.lang.psi.elements.MethodReference
import com.jetbrains.php.lang.psi.elements.ParameterList
import org.nextras.orm.intellij.utils.OrmUtils
import org.nextras.orm.intellij.utils.PhpIndexUtils

class SetValueCompletionProvider : CompletionProvider<CompletionParameters>() {
	override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
		val el = parameters.position

		if (el.parent?.parent !is ParameterList || (el.parent.parent as ParameterList).parameters[0] !== el.parent) {
			return
		}

		val methodReference = el.parent?.parent?.parent as? MethodReference ?: return
		if (methodReference.name != "setValue" && methodReference.name != "setReadOnlyValue") {
			return
		}

		val phpIndex = PhpIndex.getInstance(el.project)
		val classes = PhpIndexUtils.getByType(methodReference.classReference!!.type, phpIndex)

		classes
			.filter { OrmUtils.OrmClass.ENTITY.`is`(it, phpIndex) }
			.flatMap { it.fields }
			.filterIsInstance<PhpDocProperty>()
			.forEach {
				result.addElement(
					LookupElementBuilder.create(it.name).withTypeText(it.type.toString())
				)
			}
	}
}
